Run tests in stricter environment

tianon opened the pull request:
Use `env -i` to very explicitly control exactly which environment variables leak into our tests.  This enforces a clean separation of "build environment knobs" versus "test suite knobs".

This also includes a minor tweak to how we handle starting our integration daemon, especially to catch failure to start sooner than failing tests.

tianon commented on integration-cli/docker_utils.go:
```diff
@@ -768,6 +768,12 @@ func fakeGIT(name string, files map[string]string) (*FakeGIT, error) {
+	}
+	if output, err := exec.Command("git", "config", "user.email", "fake.user@example.com").CombinedOutput(); err != nil {
+		return nil, fmt.Errorf("error trying to set 'user.email': %s (%s)", err, output)
+	}
```
This is one of the assumptions this change caught:
```
--- FAIL: TestBuildFromGIT (0.04s)
	docker_cli_build_test.go:3973: error trying to commit to repo: exit status 128 (
		*** Please tell me who you are.
		
		Run
		
		  git config --global user.email "you@example.com"
		  git config --global user.name "Your Name"
		
		to set your account's default identity.
		Omit --global to set the identity only in this repository.
		
		fatal: unable to auto-detect email address (got 'root@9589bb267473.(none)')
		)
```

ahmetalpbalkan commented on integration-cli/docker_utils.go:
```diff
@@ -267,7 +267,7 @@ func (d *Daemon) Cmd(name string, arg ...string) (string, error) {
 
 func daemonHost() string {
 	daemonUrlStr := "unix:///var/run/docker.sock"
-	if daemonHostVar := os.Getenv("DOCKER_TEST_HOST"); daemonHostVar != "" {
```
why?

tianon replied:
https://github.com/docker/docker/blob/99ca215c65e9a482f981790f2d4dbe8d4ccce4be/project/make/.integration-daemon-start#L27

Because `DOCKER_TEST_HOST` was only ever intended to be a knob you pass to the scripts to control the `DOCKER_HOST` value used by Docker, which all these tests invoke (and only some of which actually use this `daemonHost()` value, and so are relying on the line of the script I linked above anyhow).  If those two happened to get out of sync somehow, we'd have a real fun time.

ahmetalpbalkan replied:
ok, sounds good then. I shall not make future references to DOCKER_TEST_HOST :yum:

tianon replied:
:heart:

ahmetalpbalkan commented on project/make.sh:
```diff
@@ -178,9 +178,22 @@ go_test_dir() {
+		GOPATH="$GOPATH" \
+		HOME="$DEST/fake-HOME" \
+		PATH="$PATH" \
+		TEST_DOCKERINIT_PATH="$TEST_DOCKERINIT_PATH" \
```
@tianon was gonna include CIRROS_TAR path as well #10750

ahmetalpbalkan replied:
/me realizes author is actually @tianon.
:facepalm:

tianon replied:
:smiling_imp:

tianon commented on project/make/.integration-daemon-start:
```diff
@@ -16,6 +16,7 @@ export DOCKER_GRAPHDRIVER=${DOCKER_GRAPHDRIVER:-vfs}
 export DOCKER_EXECDRIVER=${DOCKER_EXECDRIVER:-native}
 
 if [ -z "$DOCKER_TEST_HOST" ]; then
+	unset DOCKER_HOST
```
I'm wondering if we should, instead of doing this, set this to something like `unix://$DEST/docker.sock` here and pass it via `--host` to the daemon we launch just below to further insulate us from the "host" we're running this daemon inside (even though in most cases right now, that's our container).

ahmetalpbalkan replied:
Well, tests also rely on unix://var/run/docker.sock, do you mean we should always have a DOCKER_HOST to run integration-cli tests, even though it's unix://.../docker.sock?

tianon replied:
@ahmetalpbalkan we have tests that hard-code `unix://...` too?? WHAAT

tianon replied:
hrefs or it didn't happen

ahmetalpbalkan replied:
ok they're not here anymore :yum:

tianon replied:
So, this idea needs:
- https://github.com/docker/docker/pull/10747
- https://github.com/docker/docker/pull/10796
- https://github.com/docker/docker/pull/10800

And then it's viable. :+1: :heart_eyes:

jfrazelle replied:
so all those were merged :D

tianon committed:
Run tests in stricter environment

Use `env -i` to very explicitly control exactly which environment variables leak into our tests.  This enforces a clean separation of "build environment knobs" versus "test suite knobs".

This also includes a minor tweak to how we handle starting our integration daemon, especially to catch failure to start sooner than failing tests.

Signed-off-by: Andrew "Tianon" Page <admwiggin@gmail.com>

tianon commented:
Updated!  "janky" is running -- I fired off "windows" too just to make sure `env -i` works there too, and it does :+1:

tianon commented:
Arg, looks like `docker-py` has some `docker.sock` path assumptions built-in, too:
```
---> Making bundle: test-docker-py (in bundles/1.5.0-dev/test-docker-py)
+++ exec docker --daemon --debug --host unix:///go/src/github.com/docker/docker/bundles/1.5.0-dev/test-docker-py/docker.sock --storage-driver vfs --exec-driver native --pidfile /go/src/github.com/docker/docker/bundles/1.5.0-dev/test-docker-py/docker.pid
..................................E.......................E
======================================================================
ERROR: test_443 (__main__.TestRegressions)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/docker-py/tests/integration_test.py", line 1366, in test_443
    for line in self.client.build(fileobj=dfile, tag="a/b/c"):
  File "/docker-py/docker/client.py", line 476, in build
    timeout=timeout,
  File "/docker-py/docker/client.py", line 78, in _post
    return self.post(url, **self._set_request_timeout(kwargs))
  File "/usr/lib/python2.7/dist-packages/requests/sessions.py", line 497, in post
    return self.request('POST', url, data=data, **kwargs)
  File "/usr/lib/python2.7/dist-packages/requests/sessions.py", line 455, in request
    resp = self.send(prep, **send_kwargs)
  File "/usr/lib/python2.7/dist-packages/requests/sessions.py", line 558, in send
    r = adapter.send(request, **kwargs)
  File "/usr/lib/python2.7/dist-packages/requests/adapters.py", line 378, in send
    raise ConnectionError(e)
ConnectionError: UnixHTTPConnectionPool(host='localhost', port=None): Max retries exceeded with url: /run/docker.sock/v1.16/build?pull=True&nocache=False&q=False&t=a%2Fb%2Fc&forcerm=False&rm=False (Caused by <class 'socket.error'>: [Errno 2] No such file or directory)

======================================================================
ERROR: test_resource_warnings (__main__.UnixconnTestCase)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/docker-py/tests/integration_test.py", line 1347, in test_resource_warnings
    client.images()
  File "/docker-py/docker/client.py", line 648, in images
    res = self._result(self._get(self._url("/images/json"), params=params),
  File "/docker-py/docker/client.py", line 81, in _get
    return self.get(url, **self._set_request_timeout(kwargs))
  File "/usr/lib/python2.7/dist-packages/requests/sessions.py", line 467, in get
    return self.request('GET', url, **kwargs)
  File "/usr/lib/python2.7/dist-packages/requests/sessions.py", line 455, in request
    resp = self.send(prep, **send_kwargs)
  File "/usr/lib/python2.7/dist-packages/requests/sessions.py", line 558, in send
    r = adapter.send(request, **kwargs)
  File "/usr/lib/python2.7/dist-packages/requests/adapters.py", line 378, in send
    raise ConnectionError(e)
ConnectionError: UnixHTTPConnectionPool(host='localhost', port=None): Max retries exceeded with url: /run/docker.sock/v1.16/images/json?only_ids=0&all=0 (Caused by <class 'socket.error'>: [Errno 2] No such file or directory)

----------------------------------------------------------------------
Ran 59 tests in 75.097s

FAILED (errors=2)
```

tianon commented:
Submitted a PR to `docker-py` for fixing these https://github.com/docker/docker-py/pull/494.

tianon commented:
Ok, updated with a newer `docker-py` commit now that my PR there is merged, so we should be passing (it works locally -- we'll see what `janky` says). :crossed_flags:

jfrazelle commented:
LGTM

jfrazelle commented:
ping @LK4D4 or @crosbymichael

spf13 commented:
LGTM
